app/models/exercise.rb

```diff
@@ -300,35 +366,108 @@ def determine_file_role_from_proforma_file(task_node, file_node)
     return 'regular_file'
   end
 
-  def from_proforma_xml(xml_string)
-    # how to extract the proforma functionality into a different module in rails?
-    xml = Nokogiri::XML(xml_string)
-    xml.collect_namespaces
-    task_node = xml.xpath('/root/p:task')
-    description = task_node.xpath('p:description/text()')[0].content
-    self.attributes = {
-      title: task_node.xpath('p:meta-data/p:title/text()')[0].content,
-      description: description,
-      instructions: description
+  def build_proforma_xml_for_exercise_file(xml, exercise_file)
+    if exercise_file.role == 'main_file'
+      proforma_file_class = 'template'
+      comment = 'main'
+    else
+      proforma_file_class = 'internal'
+      comment = ''
+    end
+
+    xml['p'].file(exercise_file.content,
+                  'filename' => exercise_file.full_file_name,
+                  'id' => exercise_file.id,
+                  'class' => proforma_file_class,
+                  'comment' => comment
+    )
+  end
+
+  def build_proforma_xml_for_test(xml, test, index)
+    proforma = xml['p']
+    proforma.test('id' => 't' + index.to_s) {
+      proforma.title('')
+      proforma.send('test-type', 'unittest')
+      proforma.send('test-configuration') {
+        proforma.filerefs {
+          proforma.fileref('refid' => test.id.to_s)
+        }
+        xml['u'].unittest('framework' => 'JUnit 4', 'version' => '')
```

Comment:
You need to check here for which programming language the exercise is. If it is Java, then use JUnit 4 (split frame work and version), if it is Python then use Pyunit, if it is Ruby, then use RSpec. We intentionally decided to leave this hard-coded in here as everything else would require very high efforts, which do not bring us any further. So the check for the programming language is not a nice solution but the minimum viable product.